File: frontend/src/lib/api/reading_data.ts
import { apiFetch } from "./client";

export interface NoteResponse {
  id: string;
  project_paper_id: string;
  title?: string;
  content: string;
  created_at: string;
  updated_at: string;
}

export interface HighlightResponse {
  id: string;
  project_paper_id: string;
  page?: number;
  selected_text: string;
  ai_note?: string;
  created_at: string;
}

export interface ReadingProgressResponse {
  id: string;
  project_paper_id: string;
  progress_percent: number;
  last_page?: number;
  started_at?: string;
  last_opened?: string;
  completed_at?: string;
}

export interface ProjectPaperReadingData {
  project_paper_id: string;
  status?: string;
  favorite: boolean;
  priority?: number;
  added_at: string;
  notes: NoteResponse[];
  highlights: HighlightResponse[];
  reading_progress?: ReadingProgressResponse;
}

// ── Reading Data ──

export async function getReadingData(projectId: string, paperId: string): Promise<ProjectPaperReadingData> {
  return apiFetch<ProjectPaperReadingData>(`/projects/${projectId}/papers/${paperId}/reading-data`);
}

// ── Notes ──

export async function createNote(
  projectId: string,
  paperId: string,
  content: string,
  title?: string,
): Promise<NoteResponse> {
  return apiFetch<NoteResponse>(`/projects/${projectId}/papers/${paperId}/notes`, {
    method: "POST",
    body: JSON.stringify({ content, title }),
  });
}

export async function updateNote(
  projectId: string,
  paperId: string,
  noteId: string,
  content: string,
): Promise<NoteResponse> {
  return apiFetch<NoteResponse>(`/projects/${projectId}/papers/${paperId}/notes/${noteId}`, {
    method: "PATCH",
    body: JSON.stringify({ content }),
  });
}

export async function deleteNote(projectId: string, paperId: string, noteId: string): Promise<void> {
  await apiFetch(`/projects/${projectId}/papers/${paperId}/notes/${noteId}`, {
    method: "DELETE",
  });
}

// ── Highlights ──

export async function createHighlight(
  projectId: string,
  paperId: string,
  selected_text: string,
  ai_note?: string,
  page?: number,
): Promise<HighlightResponse> {
  return apiFetch<HighlightResponse>(`/projects/${projectId}/papers/${paperId}/highlights`, {
    method: "POST",
    body: JSON.stringify({ selected_text, ai_note, page }),
  });
}

export async function deleteHighlight(
  projectId: string,
  paperId: string,
  highlightId: string,
): Promise<void> {
  await apiFetch(`/projects/${projectId}/papers/${paperId}/highlights/${highlightId}`, {
    method: "DELETE",
  });
}

// ── Reading Progress ──

export async function updateReadingProgress(
  projectId: string,
  paperId: string,
  progress_percent: number,
  last_page?: number,
): Promise<ReadingProgressResponse> {
  return apiFetch<ReadingProgressResponse>(`/projects/${projectId}/papers/${paperId}/progress`, {
    method: "PATCH",
    body: JSON.stringify({ progress_percent, last_page }),
  });
}

File: backend/app/schemas/reading_data.py
import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict

# Notes
class NoteBase(BaseModel):
    title: Optional[str] = None
    content: str

class NoteCreate(NoteBase):
    pass

class NoteUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None

class NoteResponse(NoteBase):
    id: uuid.UUID
    project_paper_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

# Highlights
class HighlightBase(BaseModel):
    page: Optional[int] = None
    selected_text: str
    ai_note: Optional[str] = None

class HighlightCreate(HighlightBase):
    pass

class HighlightUpdate(BaseModel):
    page: Optional[int] = None
    selected_text: Optional[str] = None
    ai_note: Optional[str] = None

class HighlightResponse(HighlightBase):
    id: uuid.UUID
    project_paper_id: uuid.UUID
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

# Reading Progress
class ReadingProgressBase(BaseModel):
    progress_percent: int
    last_page: Optional[int] = None
    started_at: Optional[datetime] = None
    last_opened: Optional[datetime] = None
    completed_at: Optional[datetime] = None

class ReadingProgressCreate(ReadingProgressBase):
    pass

class ReadingProgressUpdate(BaseModel):
    progress_percent: Optional[int] = None
    last_page: Optional[int] = None
    started_at: Optional[datetime] = None
    last_opened: Optional[datetime] = None
    completed_at: Optional[datetime] = None

class ReadingProgressResponse(ReadingProgressBase):
    id: uuid.UUID
    project_paper_id: uuid.UUID
    
    model_config = ConfigDict(from_attributes=True)

# Aggregated Reading Data (Returned for a project context)
class ProjectPaperReadingData(BaseModel):
    project_paper_id: uuid.UUID
    status: Optional[str] = None
    favorite: bool = False
    priority: Optional[int] = None
    added_at: datetime
    notes: list[NoteResponse]
    highlights: list[HighlightResponse]
    reading_progress: Optional[ReadingProgressResponse] = None
